# Chapter 6 - Other Popular Machine Learning Models
## Segment 3- Instance-based learning w/ K-Nearest Neighbor

### Setting up for classification analysis

In [4]:
import numpy as np
import pandas as pd 
import scipy
import urllib
import sklearn

import matplotlib.pyplot as plt
from pylab import rcParams

from sklearn import neighbors
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn import metrics

In [5]:
from sklearn.neighbors import KNeighborsClassifier

In [6]:
np.set_printoptions(precision=4, suppress=True)
%matplotlib inline
rcParams['figure.figsize'] = 7,4
plt.style.use('seaborn-whitegrid')

## Importing your data

In [10]:
address = 'C:/Users/imran/OneDrive/Documents/GitHub/pythonLinkedIn/pythonWithLinkedIn/DS Essential Training 1/Data/mtcars.csv'

cars = pd.read_csv(address)

cars.columns = ['car_names', 'mpg', 'cyl', 'disp', 'hp', 'drat', 'wt', 'qsec', 'vs', 'am', 'gear', 'carb' ]
X_prime = cars[['mpg', 'disp', 'hp', 'wt']].values
y = cars.iloc[:,9].values

In [11]:
X= preprocessing.scale(X_prime)

In [12]:
X_train, X_test, y_train, y_test =train_test_split(X, y, test_size=0.2, random_state=17)

## Building and training your model with training data

In [13]:
clf = neighbors.KNeighborsClassifier()
clf.fit(X_train, y_train)
print(clf)

KNeighborsClassifier()


## Evaluating your model's predictions

In [15]:
y_pred = clf.predict(X_test)
y_expect = y_test

print(metrics.classification_report(y_expect, y_pred))

              precision    recall  f1-score   support

           0       0.80      1.00      0.89         4
           1       1.00      0.67      0.80         3

    accuracy                           0.86         7
   macro avg       0.90      0.83      0.84         7
weighted avg       0.89      0.86      0.85         7



<b> Recall: </b> a measure of model's completeness. This data is saying that out of all your points that were labeled 1, only 67% of the results were returned were truly relevant. Of the entire dataset, 83% of the results that were returned were truly relevant.
<b> High Precision + Low recall= Few results returned, but many of the label predictions that are returend are correct. </b>
<b> Recall: </b>
<b> Recall: </b>